### PR TITLE
chore: added yarn build to docker file

### PR DESCRIPTION
**Pull request by Krithika3**
In the below PR we have added the following:
1. Dockerfile: Added yarn build to the docker image and copied the UI in the main numaflow binary. (I did not research if there was a way to compress this yet)
2. The image size prior to adding UI director `ui/build` is:
Screen Shot 2022-06-28 at 8 48 37 PM
3. The image size after adding UI director `ui/build` is:
Screen Shot 2022-06-28 at 9 14 51 PM
4. The make start (which runs make image in turn runs docker-build) command with cached package.json changes and yarn build takes the following time:
Screen Shot 2022-06-28 at 10 41 58 PM
5.  I think this is reasonable and we could take a run at the release github action to see how it behaves.
6. running `kubectl apply -f ./config/install.yaml` now installs the numaflow-server too. (I have not updated the README.md yet, we could take a stab at it if this approach is ok)

### CODE REVIEW

File: config/base/numaflow-server-sa.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
```
Comment (Krithika3) [inline config/base/numaflow-server-sa.yaml:1]: Do we need a separate service account for numaflow-server? Currently I have re-used the existing one

Comment (Krithika3) [reply config/base/numaflow-server-sa.yaml:1]: What is the best practice?

File: config/cluster-install/rbac/numaflow-server-binding.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
```
Comment (Krithika3) [inline config/cluster-install/rbac/numaflow-server-binding.yaml:2]: I have added the clusterRoleBinding and the clusterRole, but re-used the existing ones. Let me know if it is good practice to add new cluster roles for server

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,16 @@
 ARG ARCH=$TARGETARCH
+####################################################################################################
+FROM docker.io/node:18 as numaflow-ui
```
Comment (Krithika3) [inline Dockerfile:3]: We can do the `yarn build` in the `make image` which would expect the user to have yarn installed, we could circumvent it by adding a flag requesting ui vs not requesting it.

Comment (Krithika3) [reply Dockerfile:3]: I think it is not warranted to do this if the docker build is not taking very long

File: config/cluster-install/rbac/numaflow-cluster-role.yaml [hunk 4]
```diff
@@ -42,6 +43,14 @@ rules:
       - update
       - patch
       - delete
+  - apiGroups:
+    - metrics.k8s.io
```
Comment (Krithika3) [inline config/cluster-install/rbac/numaflow-cluster-role.yaml:47]: Is it ok to re-use the existing cluster role?

File: Dockerfile (hunk 3 above)
Comment (whynowy) [reply Dockerfile:3]: 👍

File: config/base/numaflow-server/numaflow-server-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        app.kubernetes.io/component: numaflow-ux
+    spec:
+      serviceAccountName: numaflow-sa
+      securityContext:
+        runAsNonRoot: true
+        runAsUser: 9737
+      containers:
+        - name: numaflow-server
+          image: quay.io/numaproj/numaflow:latest
+          args: [ "server" ]
```
Comment (whynowy) [inline config/base/numaflow-server/numaflow-server-deployment.yaml:24]: Use yaml approach.

File: config/base/numaflow-server/numaflow-server-deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      labels:
+        app.kubernetes.io/part-of: numaflow
+        app.kubernetes.io/component: numaflow-ux
+    spec:
+      serviceAccountName: numaflow-sa
+      securityContext:
+        runAsNonRoot: true
+        runAsUser: 9737
+      containers:
+        - name: numaflow-server
```
Comment (whynowy) [inline config/base/numaflow-server/numaflow-server-deployment.yaml:22]: Use a name like `main` maybe?

File: config/cluster-install/rbac/numaflow-cluster-role.yaml (hunk 4 above)
Comment (whynowy) [reply config/cluster-install/rbac/numaflow-cluster-role.yaml:47]: No.

File: config/cluster-install/rbac/numaflow-server-binding.yaml (hunk 2 above)
Comment (whynowy) [reply config/cluster-install/rbac/numaflow-server-binding.yaml:2]: Please use a separate set of sa, role and binding.

File: config/base/numaflow-server/numaflow-server-deployment.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+apiVersion: apps/v1
```
Comment (whynowy) [inline config/base/numaflow-server/numaflow-server-deployment.yaml:1]: Also add a ClusterIP Service.

File: config/cluster-install/rbac/numaflow-cluster-role.yaml (hunk 4 above)
Comment (Krithika3) [reply config/cluster-install/rbac/numaflow-cluster-role.yaml]: Done

**Comment by whynowy** [comment]
I can see the time increase is about 2 mins for each image build.

File: config/base/numaflow-server/numaflow-server-service.yaml [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+kind: Service
+metadata:
+  name: numaflow-server-svc
+spec:
+  type: ClusterIP
+  ports:
+  - targetPort: 8443
+    port: 8443
+  selector:
+    app.kubernetes.io/component: numaflow-ux
```
Comment (whynowy) [inline config/base/numaflow-server/numaflow-server-service.yaml:11]: Use all the labels:

```
        app.kubernetes.io/part-of: numaflow
        app.kubernetes.io/component: numaflow-ux
```

File: config/cluster-install/rbac/numaflow-server-cluster-role.yaml [hunk 9]
```diff
@@ -0,0 +1,57 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: numaflow-server-role
+rules:
+  - apiGroups:
+      - numaflow.numaproj.io
+    verbs:
+      - create
```
[the remark this answers is not in the document]
Comment (Krithika3) [reply config/cluster-install/rbac/numaflow-server-cluster-role.yaml]: Done

**Comment by Krithika3** [comment]
> I can see the time increase is about 2 mins for each image build.

**Review by whynowy — APPROVED** [review]
LGTM

**Pull request merged**